api/include/opentelemetry/metrics/meter.h

```diff
@@ -0,0 +1,204 @@
+   * @throws IllegalArgumentException if a different metric by the same name exists in this meter.
+   * @throws IllegalArgumentException if the {@code name} does not match spec requirements.
+   */
+  virtual opentelemetry::nostd::shared_ptr<AsynchronousInstrument<T>> NewUpDownSumObserver(
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled,
+      void (*callback)(ObserverResult<T>)) = 0;
+
+  /**
+   * Creates a ValueObserver with the passed characteristics and returns a shared_ptr to that
+   * ValueObserver.
+   *
+   * @param name the name of the new ValueObserver.
+   * @param description a brief description of what the ValueObserver is used for.
+   * @param unit the unit of metric values following https://unitsofmeasure.org/ucum.html.
+   * @param enabled a boolean value that turns on or off the metric instrument.
+   * @param callback the function to be observed by the instrument.
+   * @return a shared pointer to the created ValueObserver.
+   * @throws NullPointerException if {@code name} is null
+   * @throws IllegalArgumentException if a different metric by the same name exists in this meter.
+   * @throws IllegalArgumentException if the {@code name} does not match spec requirements.
+   */
+  virtual opentelemetry::nostd::shared_ptr<AsynchronousInstrument<T>> NewValueObserver(
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled,
+      void (*callback)(ObserverResult<T>)) = 0;
+
+  /**
+   * Utility method that allows users to atomically record measurements to a set of
+   * synchronous metric instruments with a common set of labels.
+   *
+   * @param labels the set of labels to associate with this recorder.
+   * @param values a span of pairs where the first element of the pair is a metric instrument
+   * to record to, and the second element is the value to update that instrument with.
+   */
+  virtual void RecordBatch(
+      const trace::KeyValueIterable &labels,
+      const nostd::span<std::pair<nostd::shared_ptr<SynchronousInstrument<T>>,
+          T>> values) noexcept
+  {
+    // No-op
+  }
+
+  /**
+   * Utility method that allows users to atomically record measurements to a set of
+   * synchronous metric instruments with a common set of labels.
+   *
+   * This overloads the BatchRecord function to allow {@code labels} to be of type initializer list.
+   *
+   * @param labels an initializer list holding labels to associate with this recording.
+   * @param values a span of pairs where the first element of the pair is a metric instrument
+   * to record to, and the second element is the value to update that instrument with.
+   */
+  void RecordBatch(
+      std::initializer_list<std::pair<nostd::string_view, nostd::string_view>> labels,
+      const nostd::span<std::pair<nostd::shared_ptr<SynchronousInstrument<T>>,
```

Comment:
I think this doesn't allow us to have `SynchronousInstrument<int>` and `SynchronousInstrument<double>` in the same span. Is this intended?

Reply:
As of now this is intended. Each instantiation of the Meter class is only capable of creating instruments of one type. Meter<int> can create counter<int> and so on. We know that this is not ideal and we are working on a way of solving this but it involves type casting which we are weary of. Does anybody else have any suggestions on how to to make the Meter class more general?

Reply:
That's not really in line with your [design doc](https://github.com/open-telemetry/opentelemetry-cpp/pull/145/files#diff-8b4a94a083232a27cb36dc7472695f37R200). Is there a requirement to have the instruments realized via templates? Java just uses distinct methods for different types.

Thinking about how this would be used, I assume it would look like this:
```c
auto c = NewUpDownCounter<int>(...);
```
as opposed to this:
```c
auto c = NewIntUpDownCounter(...);
```

I have no strong preference and would be fine with either of those.

If you want to go with templates, you could make a universal `Meter` interface work with partial specializations. I put up a short example [here](https://gist.github.com/pyohannes/c8ef75d370dd053dc66d244c46c12e30).

Reply:
This is a very helpful example, thanks!

Sorry for the out-of-date design doc, we are actually in the process of updating it to reflect the feedback we have received. Namely, we will be templating our design. The commit introducing that to the design doc will be coming very soon :)